File: A2Z_striver/patterns/pattern_20/pattern_20.py
import sys

class Solution:
    """
    Problem: Print Butterfly Pattern (Single Peak).
    Example Input: n = 3
    Example Output:
    * *
    ** **
    ****** (Row n-1: Peak)
    ** **
    * *
    Goal: O(N) Time (due to string mult) or O(N^2) characters.
    """

    # // ~ Approach 1: Pythonic Split Loops (Your Code)
    # // Analysis: Time O(N) lines | Space O(N) string construction
    # // Pros: Extremely clean, leverages Python's string multiplication.
    def pattern20_split(self, n):
        # Part 1: Growing Phase (0 to n-1)
        for i in range(n):
            stars_count = i + 1
            spaces_count = 2 * n - 2 * stars_count
            
            # String Multiplication: '*' * 3 -> "***"
            stars = "*" * stars_count
            spaces = " " * spaces_count
            print(stars + spaces + stars)
            
        # Part 2: Shrinking Phase (0 to n-2)
        # We run n-1 times to avoid duplicating the middle row
        for i in range(n - 1):
            stars_count = n - 1 - i
            spaces_count = 2 * n - 2 * stars_count
            
            stars = "*" * stars_count
            spaces = " " * spaces_count
            print(stars + spaces + stars)

    # // ~ Approach 2: Single Loop (Coordinate Math)
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Universal logic (works in C++/Java too), demonstrates index mastery.
    def pattern20_single_loop(self, n):
        # Run for 2*n - 1 rows (Single Peak variant)
        # Indices: 0, 1, ... n-1 (Peak) ... 2n-2
        for i in range(2 * n - 1):
            
            # // --- 1. MIRROR LOGIC ---
            # // Goal: Map bottom indices back to top indices.
            # // Top Half (i < n): row = i
            # // Bottom Half (i >= n):
            # //    - Last Row Index is (2*n - 2). We want this to map to Row 0.
            # //    - Formula: row = (LastIndex) - i
            # //    - Formula: row = (2*n - 2) - i
            row = i
            if i >= n:
                row = 2 * n - 2 - i
            
            # // --- 2. BOUNDARY LOGIC ---
            # // We need stars on the edges, spaces in the middle.
            # // Left Star Count: row + 1 (Row 0 has 1 star)
            # // Right Star Start: TotalWidth - LeftStars
            
            stars_limit = row + 1
            total_width = 2 * n
            
            for j in range(total_width):
                # // Condition:
                # // Left Side: j < stars_limit
                # // Right Side: j >= total_width - stars_limit
                if j < stars_limit or j >= total_width - stars_limit:
                    print("*", end="")
                else:
                    print(" ", end="")
            print()

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    n = 3
    print(f"--- Approach 1: Pythonic Split (n={n}) ---")
    solver.pattern20_split(n)
    
    print(f"\n--- Approach 2: Single Loop (n={n}) ---")
    solver.pattern20_single_loop(n)

File: A2Z_striver/patterns/pattern_20/pattern_20.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print Butterfly Pattern (Single Peak).
// Example Input: n = 3
// Example Output:
// *....* (Row 0: 1 star each side)
// **..** (Row 1: 2 stars each side)
// ****** (Row 2: 3 stars each side - PEAK)
// **..** (Row 3: Mirror of Row 1)
// *....* (Row 4: Mirror of Row 0)
// Goal: O(N^2) Time.

class Solution
{
public:
    // ~ Approach 1: Single Loop with Mirror Math (Your Approach)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Short code, demonstrates index mapping skills.
    void pattern20_single_loop(int n)
    {
        // Run for 2*n - 1 rows (Single Peak variant)
        for (int i = 0; i < 2 * n - 1; i++)
        {

            // --- 1. MIRROR LOGIC ---
            // We need an 'effective_row' index that goes: 0 -> 1 -> 2 -> 1 -> 0
            // Top Half (i < n): effective_row = i
            // Bottom Half (i >= n):
            //    Formula derivation:
            //    - Total Height = 2*n - 1. Last Index = 2*n - 2.
            //    - We want Last Index to map to 0.
            //    - as, Last Index - i == 0 true.
            //    - So, effective_row = (2*n - 2) - i

            int row = i;
            if (i >= n)
            {
                row = 2 * n - 2 - i;
            }

            // --- 2. INNER LOOP ---
            // Total Width is 2*n.
            // Left Stars: indices 0 to 'row' (count = row+1)
            // Right Stars: indices starting from (2*n - 1 - row) to end.

            for (int j = 0; j < 2 * n; j++)
            {

                // Condition:
                // j <= row           -> Left Triangle
                // j >= total_width(i.e the last column) - row -> Right Triangle
                // j >= 2*n - 1 - row -> Right Triangle
                if (j <= row || j >= 2 * n - 1 - row)
                {
                    cout << "*";
                }
                else
                {
                    cout << " ";
                }
            }
            cout << "\n";
        }
    }

    // ~ Approach 2: Split Loops (Iterative / Readable)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Easier to debug, no complex formulas.
    void pattern20_split(int n)
    {
        // Part 1: Growing Phase (0 to n-1) -- Includes the Peak row
        for (int i = 0; i < n; i++)
        {
            int stars = i + 1;
            int spaces = 2 * n - 2 * stars;

            // Left Stars
            for (int j = 0; j < stars; j++)
                cout << "*";
            // Spaces
            for (int j = 0; j < spaces; j++)
                cout << " ";
            // Right Stars
            for (int j = 0; j < stars; j++)
                cout << "*";

            cout << "\n";
        }

        // Part 2: Shrinking Phase (n-2 down to 0) -- Skips Peak row
        // Note: Loop starts from n-2 to avoid duplicating the middle row
        for (int i = n - 2; i >= 0; i--)
        {
            int stars = i + 1;
            int spaces = 2 * n - 2 * stars;

            // Left Stars
            for (int j = 0; j < stars; j++)
                cout << "*";
            // Spaces
            for (int j = 0; j < spaces; j++)
                cout << " ";
            // Right Stars
            for (int j = 0; j < stars; j++)
                cout << "*";

            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 3;

    cout << "--- Approach 1: Single Loop (Math) ---\n";
    sol.pattern20_single_loop(n);

    cout << "\n--- Approach 2: Split Loops (Iterative) ---\n";
    sol.pattern20_split(n);

    return 0;
}